### PR TITLE
feat: AsyncIO Integration [Part 2]

### PR DESCRIPTION
**Pull request by lidizheng**
Children PR of https://github.com/googleapis/python-api-core/pull/26.

This PR includes AsyncIO version of:
* Polling future
* Page iterator

The AsyncIO version of polling future still uses the same mechanism as the sync version. The AsyncIO polling future tries to update its own state whenever the application want to access information or perform actions.

For page iterator, it has similar interface design as sync version. But instead of fulfilling normal generator protocol, it is based on the async generator.

Related #23

### CODE REVIEW

File: google/api_core/operation_async.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    def _set_result_from_operation(self):
+        """Set the result or exception from the operation if it is complete."""
+        # This must be done in a lock to prevent the async_future thread
+        # and main thread from both executing the completion logic
+        # at the same time.
+        with self._completion_lock:
+            # If the operation isn't complete or if the result has already been
+            # set, do not call set_result/set_exception again.
+            # Note: self._result_set is set to True in set_result and
+            # set_exception, in case those methods are invoked directly.
```
Comment (busunkim96) [inline google/api_core/operation_async.py:124]: Stale comment?

File: google/api_core/operation_async.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+    operation = my_api_client.long_running_method()
+
+    def my_callback(future):
+        result = future.result()
+
+    operation.add_done_callback(my_callback)
+
+"""
+
```
Comment (busunkim96) [inline google/api_core/operation_async.py:38]: Should these snippets use `await`?

File: google/api_core/future/async_future.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        If the operation is completed, the callback will be scheduled onto the
+        event loop. Otherwise, the callback will be stored and invoked when the
+        future is done.
+
+        Args:
+            fn (Callable[Future]): The callback to execute when the operation
+                is complete.
+        """
+        if self._background_task is None:
+            self._background_task = asyncio.get_event_loop().create_task(self._blocking_poll())
```
Comment (busunkim96) [inline google/api_core/future/async_future.py:148]: If I'm reading [this page](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task) correctly, `asyncio.create_task` is only available in 3.7+?

File: docs/operation.rst [hunk 4]
```diff
@@ -4,3 +4,10 @@ Long-Running Operations
 .. automodule:: google.api_core.operation
   :members:
   :show-inheritance:
+
+LRO in AsyncIO
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply docs/operation.rst:8]: Done.

File: google/api_core/future/async_future.py (hunk 3 above)
Comment (lidizheng) [reply google/api_core/future/async_future.py:148]: Yes. `asyncio.create_task()` is 3.7+, but [`loop.create_task()`](https://docs.python.org/3.8/library/asyncio-eventloop.html#asyncio.loop.create_task) is 3.5+.

File: google/api_core/operation_async.py (hunk 2 above)
Comment (lidizheng) [reply google/api_core/operation_async.py:38]: Good catch! I missed this docstring. Updated to:
```
"""AsyncIO futures for long-running operations returned from Google Cloud APIs.

These futures can be used to await for the result of a long-running operation
using :meth:`AsyncOperation.result`:


.. code-block:: python

    operation = my_api_client.long_running_method()
    result = await operation.result()

Or asynchronously using callbacks and :meth:`Operation.add_done_callback`:

.. code-block:: python

    operation = my_api_client.long_running_method()

    def my_callback(future):
        result = await future.result()

    operation.add_done_callback(my_callback)

"""
```

File: google/api_core/operation_async.py (hunk 1 above)
Comment (lidizheng) [reply google/api_core/operation_async.py:124]: This note no longer apply to AsyncIO, removed.

File: docs/operation.rst [hunk 5]
```diff
@@ -4,3 +4,10 @@ Long-Running Operations
 .. automodule:: google.api_core.operation
   :members:
   :show-inheritance:
+
+Long-Running Operations in AsyncIO
+--------------
```
Comment (busunkim96) [inline docs/operation.rst:9]: (To fix the docs build)
```suggestion
-------------------------------------
```

Comment (lidizheng) [reply docs/operation.rst]: One of reasons why Markdown is more popular than RST...

File: google/api_core/page_iterator_async.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        """
+        if self.page_number == 0:
+            return True
+
+        if self.max_results is not None:
+            if self.num_results >= self.max_results:
+                return False
+
+        # Note: intentionally a falsy check instead of a None check. The RPC
+        # can return an empty string indicating no more pages.
```
Comment (software-dov) [inline google/api_core/page_iterator_async.py:277]: Nit: move this comment to right above the 
```python
if self.max_results is not None
```

**Comment by lidizheng** [comment]
Thanks for reviewing 👍

**Pull request merged**